packaging/standalone/src/main/distribution/shell-scripts/bin/neo4j-admin

```diff
@@ -1,44 +1,193 @@
+  detect_os
+  _find_java_home
+
+  if [[ "${JAVA_HOME:-}" ]] ; then
+    JAVA_CMD="${JAVA_HOME}/bin/java"
+    if [[ ! -f "${JAVA_CMD}" ]]; then
+      echo "ERROR: JAVA_HOME is incorrectly defined as ${JAVA_HOME} (the executable ${JAVA_CMD} does not exist)"
+      exit 1
+    fi
+  else
+    if [ "${DIST_OS}" != "macosx" ] ; then
+      # Don't use default java on Darwin because it displays a misleading dialog box
+      JAVA_CMD="$(which java || true)"
+    fi
+  fi
+
+  if [[ ! "${JAVA_CMD:-}" ]]; then
+    echo "ERROR: Unable to find Java executable."
+    _show_java_help
+    exit 1
+  fi
+}
+
+_find_java_home() {
+  [[ "${JAVA_HOME:-}" ]] && return
+
+  case "${DIST_OS}" in
+    "macosx")
+      JAVA_HOME="$(/usr/libexec/java_home -v 1.8)"
+      ;;
+    "gentoo")
+      JAVA_HOME="$(java-config --jre-home)"
+      ;;
+  esac
+}
+
+_show_java_help() {
+  echo "* Please use Oracle(R) Java(TM) 8, OpenJDK(TM) or IBM J9 to run Neo4j."
+  echo "* Please see https://neo4j.com/docs/ for Neo4j installation instructions."
+}
+
+_setup_calculated_paths() {
+  if [[ -z "${NEO4J_HOME:-}" ]]; then
+    NEO4J_HOME="$(cd "$(dirname "$0")"/.. && pwd)"
+  fi
+  : "${NEO4J_CONF:="${NEO4J_HOME}/conf"}"
+  readonly NEO4J_HOME NEO4J_CONF
+}
+
+_read_config() {
+  # - plain key-value pairs become environment variables
+  # - keys have '.' chars changed to '_'
+  # - keys of the form KEY.# (where # is a number) are concatenated into a single environment variable named KEY
+  parse_line() {
+    line="$1"
+    if [[ "${line}" =~ ^([^#\s][^=]+)=(.+)$ ]]; then
+      key="${BASH_REMATCH[1]//./_}"
+      value="${BASH_REMATCH[2]}"
+      if [[ "${key}" =~ ^(.*)_([0-9]+)$ ]]; then
+        key="${BASH_REMATCH[1]}"
+      fi
+      if [[ "${!key:-}" ]]; then
+        export ${key}="${!key} ${value}"
+      else
+        export ${key}="${value}"
+      fi
+    fi
+  }
+
+  if [[ -f "${NEO4J_CONF}/neo4j-wrapper.conf" ]]; then
+    cat >&2 <<EOF
+WARNING: neo4j-wrapper.conf is deprecated and support for it will be removed in a future
+         version of Neo4j; please move all your settings to neo4j.conf
+EOF
+  fi
+
+  for file in "neo4j-wrapper.conf" "neo4j.conf"; do
+    path="${NEO4J_CONF}/${file}"
+    if [ -e "${path}" ]; then
+      while read line; do
+        parse_line "${line}"
+      done <"${path}"
+    fi
+  done
 }
 
-main "$@"
+_setup_configurable_paths() {
+  NEO4J_DATA=$(resolve_path "${dbms_directories_data:-data}")
+  NEO4J_LIB=$(resolve_path "${dbms_directories_lib:-lib}")
+  NEO4J_LOGS=$(resolve_path "${dbms_directories_logs:-logs}")
+  NEO4J_PLUGINS=$(resolve_path "${dbms_directories_plugins:-plugins}")
+  NEO4J_RUN=$(resolve_path "${dbms_directories_run:-run}")
+  readonly NEO4J_DATA NEO4J_LIB NEO4J_LOGS NEO4J_PLUGINS NEO4J_RUN
+}
+
+
+setup_heap() {
+  JAVA_MEMORY_OPTS=()
+  if [[ -n "${HEAP_SIZE:-}" ]]; then
+    JAVA_MEMORY_OPTS+=("-Xmx${HEAP_SIZE}")
+    JAVA_MEMORY_OPTS+=("-Xms${HEAP_SIZE}")
+  fi
+}
+setup_heap
```

Comment:
only neo4j-admin should call this particular heap logic so it can't go in the general call_main_class function though.